apps/full-site-editing/full-site-editing-plugin/full-site-editing/class-full-site-editing.php

```diff
@@ -350,4 +355,128 @@ public function allow_searching_for_templates() {
 		// Setting this to `public` will allow it to be found in the search endpoint.
 		$post_type->public = true;
 	}
+
+	/**
+	 * This will merge the post content with the post template, modifiying the
+	 * $post parameter.
+	 *
+	 * @param WP_Post $post
+	 */
+	public function merge_template_and_post( $post ) {
+		//bail if not a REST API Request
+		if ( defined( 'REST_REQUEST' ) && ! REST_REQUEST ) {
+			return;
+		}
+
+		// bail if the post type is one of the template post types
+		if ( in_array( get_post_type( $post->ID ), $this->template_post_types ) ) {
+			return;
+		}
+
+		$template_id = get_post_meta( $post->ID, '_wp_template_id', true );
+		//bail if the post has no tempalte id assigned
+		if ( ! $template_id ) {
+			return;
+		}
+
+		$template = get_post( $template_id );
+		$wrapped_post_content = sprintf( '<!-- wp:a8c/post-content -->%s<!-- /wp:a8c/post-content -->', $post->post_content );
+		$post->post_content = str_replace( '<!-- wp:a8c/post-content /-->', $wrapped_post_content, $template->post_content );
+	}
+
+	/**
+	 * This will extract the inner blocks of the post content and
+	 * serialize them back to HTML for saving.
+	 *
+     * @param array $data    An array of slashed post data.
+	 * @param array $postarr An array of sanitized, but otherwise unmodified post data.
+	 */
+	public function remove_template_components( $data, $postarr ) {
+		// bail if the post type is one of the template post types
+		if ( in_array( $postarr['post_type'], $this->template_post_types ) ) {
+			return $data;
+		}
+
+		$post_content = wp_unslash( $data['post_content'] );
+
+		//bail if post content has no blocks
+		if( ! has_blocks( $post_content ) ) {
+			return $data;
+		}
+
+		$post_content_blocks = parse_blocks( wp_unslash( $data['post_content'] ) );
```

Comment:
Since we've already extracted it in a separate variable bellow we could avoid another unslash call:
```suggestion
		$post_content_blocks = parse_blocks( $post_content );
```